pilot/cmd/pilot-agent/options/security.go

```diff
@@ -65,8 +71,20 @@ func NewSecurityOptions(proxyConfig *meshconfig.ProxyConfig, stsPort int, tokenM
 	var tokenManager security.TokenManager
 	if stsPort > 0 || xdsAuthProvider.Get() != "" {
 		// tokenManager is gcp token manager when using the default token manager plugin.
-		tokenManager = tokenmanager.CreateTokenManager(tokenManagerPlugin,
-			tokenmanager.Config{CredFetcher: o.CredFetcher, TrustDomain: o.TrustDomain})
+		for retryCount := 1; retryCount <= tokenManagerMaxNumRetries; retryCount++ {
+			tokenManager, err = tokenmanager.CreateTokenManager(tokenManagerPlugin,
+				tokenmanager.Config{CredFetcher: o.CredFetcher, TrustDomain: o.TrustDomain})
+			if err != nil {
+				if retryCount < tokenManagerMaxNumRetries {
+					log.Errorf("failed to create token manager: %v, retry %v ...", err, retryCount)
```

Comment:
I don't think a retry helps. The GCP metadata fetch is cached, if it fails once we would reuse the empty result.

Also I think there are two different STS token managers - probably both has the same issue.

Reply:
- Based on the comment, the retry mechanism is removed.
- The error message “token manager failed to generate access token: no plugin is found” indicates that it results from CreateTokenManager(). If other token managers encounter the same error messages, we can address them in a follow-up PR.

Reply:
That does not make sense to me. We can be 100% certain if the bug exists there it also exists here: https://github.com/istio/istio/blob/562cdb883aa613efdf6c2584709f6a1f7adf99a0/pilot/cmd/pilot-agent/options/security.go#L116 because its calling the exact same code.

If we do tests with GCE metadata server degraded we will quickly find ^ is broken as well. 

The error logs from the bug reports show the code above also has issues.

Reply:
The code has been updated to:
- Add error detection to constructAudience().
- When NewSecureTokenServiceExchanger() fails, return the error.